### ISSUE TITLE
Inactive variables in spir-v

### ISSUE THREAD

**Issue by hiddenbitious** [KhronosGroup/glslang#482]
Inactive variables (varying inputs/outputs, ubos etc) although correctly omitted from the shader reflection, they do appear in the generated spir-v. Is there a (simple) way to eliminate them from the final spir-v?

**Comment by johnkslang** [comment]
There tooling ecosystem design here is to have independent (modular) tools that do SPIR-V transforms, which wolud include dead-code elimination and related.  A collection is being started in https://github.com/KhronosGroup/SPIRV-Tools, the remapper does some things and possible also should move there, and there may be others people are working on.

Having said that, one should be careful in changing the interface to a SPIR-V stage, as that could change a matching interface to a non-matching interface.

**Comment by hiddenbitious** [comment]
So if I understand this correctly, eliminating shader's interface inactive variables (and possibly other inactive code) depends on the entry point, but detecting all possible situations is not in glslang's responsibilities.

**Comment by johnkslang** [comment]
Right.  Glslang is not trying to be the only SPIR-V tool that does everything possible with SPIR-V.  It's primary object in this space is correct GLSL -> SPIR-V, rather than creating optimal/smallest/combined/specialized/etc. SPIR-V.  The repository also contains the remapper, which has useful SPIR-V transforms, which is counter to this a but, as logically the remapper belongs more with the SPIR-V tools repository.

It depends on the tool chain how a set of modular transforms might be hooked up to create what it wants to ship:
 - What shaders/stages should all go into the same SPIR-V module?
 - When/where/whether DCE should be done
 - What set of modules should be remapped and/or compressed together into a single shipping unit?
 - etc.

**Comment by hiddenbitious** [comment]
Ok this makes sense. Thanks.

**Comment by hiddenbitious** [comment]
Sorry for bringing this back. I used the remapper (tried various command line argument combinations --dce all, --do-everything etc) on a shader but the inactive variables (just one, tried a trivial shader) still appear in the final spir-v. Am I doing something wrong or is this a known issue?

**Comment by johnkslang** [comment]
It's pretty dangerous for a tool to change interfaces, so the remapper would be cautious about that.  Sorry if I implied the remapper shrank interfaces.  Perhaps a new tool could, or a new mode on the remapper could.

But, I would want to understand better why you want a tool to modify interfaces before recommending it.

**Comment by hiddenbitious** [comment]
Actually the spv::spirvbin_base_t::DCE_VARS gave me that impression and decided to try it. 

To be precise I don't want to modify shaders' interface but I want to know the active inputs/outputs (as well as uniform blocks) in order to connect shaders with the fixed-function segments of a rendering pipeline.

I am using SPIRV-Cross to parse the spir-v which (from today actually) can detect inactive variables and discard them from the glsl/c++ it generates. So my problem, for now at least, seems to have been solved.

**Comment by eugeneko** [comment]
I have somewhat related issue. I wonder if anyone here have any advice.

Output SPIRV has inactive vertex inputs, and it's 100% fine for me -- I don't need any optimization and I don't want to link against SPIRV-Tools.

However, how am I supposed to remap inactive inputs if `glslang` doesn't give _any_ information about what these inputs are?
Current reflection API is completely worthless for me. I have `getNumLiveAttributes`, but it has nothing to do with actual attrbutes in generated SPIRV bytecode.

Basically, `program.buildReflection` is _not_ a reflection of `glslang::GlslangToSpv` bytecode.
Is it API bug? Or maybe I'm doing something wrong?
